cylc/flow/task_pool.py

```diff
@@ -498,16 +432,126 @@ def release_runahead_task(self, itask):
+                # Auto-spawn next instance of tasks with no parents at the next
+                # point (or with all parents before the suite start point).
+                self.spawn(itask.tdef.name, itask.point,
+                           itask.tdef.name, next_point)
+            else:
+                # Autos-spawn next instance of tasks with absolute triggers.
+                abs_triggers = itask.tdef.get_abs_triggers(next_point)
+                for trig in abs_triggers:
+                    self.spawn(trig.task_name, trig.get_point(next_point),
+                               itask.tdef.name, next_point, trig.output,
+                               finished=True)
+
+    def _get_oldest_active_point(self):
+        """Return oldest point with active tasks, or None.
+
+        Used for removing succeeded tasks where alternate paths join (these
+        have at least one parent that will never be satisfied). Active includes
+        anything but succeeded.
+
+        """
+        active_points = []
+        for point, itasks in sorted(self.get_main_tasks_by_point().items()):
+            for itask in itasks:
+                if not itask.state(TASK_STATUS_SUCCEEDED):
+                    active_points.append(point)
+                    break
+        if active_points:
+            return active_points[0]
+        else:
+            return None
+
+    def housekeep_tasks(self):
+        """Remove spent task proxies.
+
+        "Finished" = succeeded, or failed so long as :fail is handled.
+
+        A | B => C
+
+        Finished tasks can be removed if all of their parents are finished (any
+        earlier could result in reflow, if a parent executes later).
+
+        For finished tasks whose parents do not finish (A and B might be on
+        alternate paths!) we use cycle point housekeeping: remove if the
+        active pool has move on past the point that could trigger the parent.
+
+        A & B => C
+
+        Waiting tasks can be removed if all of their parents have finished (at
+        which point nothing can happen to satisfy them). (Except for tasks with
+        no prerequisites and/or waiting on xtriggers).
+
+        Unhandled failed tasks and their downstream waiting children are kept
+        in the pool for two reasons:
+        - to keep the n=0 window focused on failed tasks
+        - the waiting children retain the state of any other prerequisites so
+          they are ready to go if the parent gets successfully retriggered.
+          (otherwise the other prerequisites belong to the "previous flow").
+
+        """
+        oldest_active_point = self._get_oldest_active_point()
+        if oldest_active_point is None:
+            # No active tasks
+            # Keep the most recent active tasks in the pool.
+            LOG.warning("Workflow stalled!")
+            return False
+
+        # TODO - adjust oldest_active_point for intercycle and future triggers!
```

Comment:
That's actually a lesser version of what we have to do already in SoS, to know when to clean up finished tasks.  In SoD it only matters for housekeeping of (the few) succeeded tasks that are kept around because they have unfinished parents.